scipy/stats/_rvs_sampling.py

```diff
@@ -168,3 +171,373 @@ def rvs_ratio_uniforms(pdf, umax, vmin, vmax, size=1, c=0, random_state=None):
+    that approximate `dist.ppf` and `dist.rvs`. For some distributions,
+    these methods may be faster than those of `dist` itself.
+
+    Parameters
+    ----------
+    dist : scipy.stats.rv_frozen
+        Frozen distribution for which fast approximate PPF is desired.
+    tol : float, optional
+        u-error tolerance. The default is 1e-12.
+    max_intervals : int, optional
+        Maximum number of intervals in the cubic Hermite Spline used to
+        approximate the percent point function. The default is 100000.
+
+    Returns
+    -------
+    H: scipy.interpolate.CubicHermiteSpline
+        Interpolant of the distributions's PPF.
+    intervals : int
+        The number of intervals of the interpolant.
+    midpoint_error : float
+        The maximum u-error at an interpolant interval midpoint.
+
+    """
+    dist, tol, max_intervals = _fni_input_validation(dist, tol, max_intervals)
+
+    # [1] Section 2.1: "For distributions with unbounded domain, we have to
+    # chop off its tails at [a] and [b] such that F(a) and 1-F(b) are small
+    # compared to the maximal tolerated approximation error."
+    p = np.array([dist.ppf(tol/10), dist.isf(tol/10)])  # initial interval
+
+    # [1] Section 2.3: "We then halve this interval recursively until
+    # |u[i+1]-u[i]| is smaller than some threshold value, for example, 0.05."
+    u = dist.cdf(p)
+    while p.size-1 <= np.ceil(max_intervals/2):
+        i = np.nonzero(np.diff(u) > 0.05)[0]
+        if not i.size:
+            break
+
+        p_mid = (p[i] + p[i+1])/2
+        # Compute only the new values and insert them in the right places
+        # [1] uses a linked list; we can't do that efficiently
+        u_mid = dist.cdf(p_mid)
+        p = np.concatenate((p, p_mid))
+        u = np.concatenate((u, u_mid))
+        i_sort = np.argsort(p)
+        p = p[i_sort]
+        u = u[i_sort]
+
+    # [1] Section 2.3: "Now we continue with checking the error estimate in
+    # each of the intervals and continue with splitting them until [it] is
+    # smaller than a given error bound."
+    u = dist.cdf(p)
+    f = dist.pdf(p)
+    while p.size-1 <= max_intervals:
+        # [1] Equation 4-8
+        H = CubicHermiteSpline(u, p, 1/f)
+        # To improve performance, add update feature to CubicHermiteSpline
+
+        # [1] Equation 12
+        u_mid = (u[:-1] + u[1:])/2
+        eu = np.abs(dist.cdf(H(u_mid)) - u_mid)
+
+        i = np.nonzero(eu > tol)[0]
+        if not i.size:
+            break
+
+        p_mid = (p[i] + p[i+1])/2
+        u_mid = dist.cdf(p_mid)
+        f_mid = dist.pdf(p_mid)
+        p = np.concatenate((p, p_mid))
+        u = np.concatenate((u, u_mid))
+        f = np.concatenate((f, f_mid))
+        i_sort = np.argsort(p)
+        p = p[i_sort]
+        u = u[i_sort]
+        f = f[i_sort]
+
+    # todo: add test for monotonicity [1] Section 2.4
+    # todo: deal with vanishing density [1] Section 2.5
```

Comment:
Still needed, planned?

Reply:
Besides accuracy, monotonicity is a desirable property of an approximate inverse of a CDF. Currently, there is a possibility that the Hermite spline is not monotonic _between_ adjacent mesh points (even if it is sufficiently _accurate_). The paper has a way to address that possibility by checking for the condition and refining the mesh in that area. It didn't seem to be as important to me as the main algorithm, so i left it out for simplicity. I'll let reviewers decide how important it is for an initial release, or whether it can be left for improvement. .

The second part would be to allow users to identify points between the leftmost and rightmost ends of the distribution support where the PDF vanishes, and this would avoid an error where the derivative of the Hermite spline is infinite. Although there could be user-defined distributions with this property, I don't know if any such distributions in SciPy, so this also didn't seem as high priority to me. But again, I'll let reviewers decide whether to require it now or leave it for improvement.

Reply:
Thanks for the explanations. I am not suited to help with this I am afraid, so I will let someone else decide.

Reply:
AFAIK, UNU.RAN checks for monotonicity. I am also not aware how often would one run into the problem. Maybe this can be taken up in a follow-up PR?